### PR TITLE
RavenDB-14893 Ensure we have enough bits for new value

### PR DESCRIPTION
**Pull request by karmeli87**

### CODE REVIEW

**Review by ayende — COMMENTED** [review]
Why do we need this?
The whole point of the temporary buffer was that we can do that check once, at the end.
As it stands now, we have to do a lot of branches in this code, which I would rather avoid.

**Comment by karmeli87** [comment]
Okay, will throw a dedicated exception for that.

File: src/Raven.Server/Documents/TimeSeries/BitsBuffer.cs [hunk 1]
```diff
@@ -199,7 +199,8 @@ public void SetBits(int bitsPosition, ulong value, int bitsInValue)
 
         public void AddValue(ulong value, int bitsInValue)
         {
-            Debug.Assert(HasEnoughBits(Header->UncompressedBitsPosition, bitsInValue));
+            if (HasEnoughBits(Header->UncompressedBitsPosition, bitsInValue) == false)
+                NotEnoughBitsException.Throw("Not enough space in the buffer");
```
Comment (ayende) [inline src/Raven.Server/Documents/TimeSeries/BitsBuffer.cs]: Were you able to get this to happen?

File: src/Raven.Server/Documents/TimeSeries/TimeSeriesValuesSegment.cs [hunk 2]
```diff
@@ -126,110 +126,119 @@ public static void AssertValueStatus(ulong status)
...
+                            return false;
+                        }
                     }
-                }
 
-                Memory.Copy(Header, tempHeader, copiedHeaderSize);
+                    Memory.Copy(Header, tempHeader, copiedHeaderSize);
 
-                return true;
+                    return true;
```
Comment (ayende) [inline src/Raven.Server/Documents/TimeSeries/TimeSeriesValuesSegment.cs]: Look at lines 216 here, we are doing explicit checks on the temp buffer, so this shouldn't be the case.
How can you get this error?

File: src/Raven.Server/Documents/TimeSeries/BitsBuffer.cs (hunk 1 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/TimeSeries/BitsBuffer.cs]: Yes, you can run the test.

File: src/Raven.Server/Documents/TimeSeries/TimeSeriesValuesSegment.cs (hunk 2 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/TimeSeries/TimeSeriesValuesSegment.cs]: This check is too late... either the assert kills me or an `ArgumentException`/`IndexOutOfRangeException` when I'm trying to add to the `tempBitBuffer` when there is not space left.

**Review by ayende — COMMENTED** [review]
Let's talk about it, I think that this is the wrong way to go about it.

File: src/Raven.Server/Documents/TimeSeries/TimeSeriesValuesSegment.cs [hunk 3]
```diff
@@ -139,7 +139,8 @@ public bool Append(ByteStringContext allocator, int deltaFromStart, Span<double>
                 sizeof(int) + // max timestamp
                 1 + // timestamp uses ControlValue which takes 4 bits, need to account for this
                 sizeof(double) * vals.Length + vals.Length /* may use additional 2 bits per value here for the first items */ +
-                2 // previous tag position (10 bits)
+                2 + // previous tag position (10 bits)
+                1 // take into account the status
```
Comment (ayende) [inline src/Raven.Server/Documents/TimeSeries/TimeSeriesValuesSegment.cs]:
```suggestion
                1 // take into account the status bit
```

**Review by ppekrol — APPROVED** [review]

**Pull request merged**